src/sentry/search/snuba/executors.py

```diff
@@ -71,134 +40,184 @@ def get_search_filter(search_filters, name, operator):
-    for h in having:
-        alias = h[0]
-        required_aggregations.add(alias)
-
-    aggregations = []
-    for alias in required_aggregations:
-        aggregations.append(aggregation_defs[alias] + [alias])
-
-    if cursor is not None:
-        having.append((sort_field, ">=" if cursor.is_prev else "<=", cursor.value))
-
-    selected_columns = []
-    if get_sample:
-        query_hash = md5(repr(conditions)).hexdigest()[:8]
-        selected_columns.append(("cityHash64", ("'{}'".format(query_hash), "issue"), "sample"))
-        sort_field = "sample"
-        orderby = [sort_field]
-        referrer = "search_sample"
-    else:
-        # Get the top matching groups by score, i.e. the actual search results
-        # in the order that we want them.
-        orderby = ["-{}".format(sort_field), "issue"]  # ensure stable sort within the same score
-        referrer = "search"
-
-    snuba_results = snuba.dataset_query(
-        dataset=Dataset.Events,
-        start=start,
-        end=end,
-        selected_columns=selected_columns,
-        groupby=["issue"],
-        conditions=conditions,
-        having=having,
-        filter_keys=filters,
-        aggregations=aggregations,
-        orderby=orderby,
-        referrer=referrer,
-        limit=limit,
-        offset=offset,
-        totals=True,  # Needs to have totals_mode=after_having_exclusive so we get groups matching HAVING only
-        turbo=get_sample,  # Turn off FINAL when in sampling mode
-        sample=1,  # Don't use clickhouse sampling, even when in turbo mode.
-    )
-    rows = snuba_results["data"]
-    total = snuba_results["totals"]["total"]
+        if environment_ids is not None:
+            filters["environment"] = environment_ids
 
-    if not get_sample:
-        metrics.timing("snuba.search.num_result_groups", len(rows))
+        if group_ids:
+            filters[self.TABLE_ALIAS + "issue"] = sorted(group_ids)
 
-    return [(row["issue"], row[sort_field]) for row in rows], total
+        conditions = []
+        having = []
+        for search_filter in search_filters:
+            if (
+                # Don't filter on issue fields here, they're not available
+                search_filter.key.name in self.issue_only_fields
+                or
+                # We special case date
+                search_filter.key.name == "date"
+            ):
+                continue
+            converted_filter = convert_search_filter_to_snuba_query(search_filter)
+            converted_filter, converted_name = self._transform_converted_filter(
+                search_filter, converted_filter, project_ids, environment_ids
+            )
+            # field_name = self.TABLE_ALIAS + search_filter.key.name
+            # Ensure that no user-generated tags that clashes with aggregation_defs is added to having
+            if converted_name in self.aggregation_defs and not search_filter.key.is_tag:
```

Comment:
Anything the parent class depends on that is supposed to be implemented by the subclasses should be declared in the parent class as abstract, so that python will refuse to instantiate an implementation that is missing that.

Reply:
```
@abstractproperty
    def aggregation_defs(self):
        """This method should return a dict of key:value
        where key is a field name for your aggregation
        and value is the aggregation function"""
        raise NotImplementedError
```